File: src/services/adPricing.ts
// Advertising rate card. One honest formula, shared by the server (which sets
// the authoritative amount an advertiser is charged) and the client (which
// shows a live estimate as the advertiser drags the reach slider). Keeping the
// maths here — and mirrored byte-for-byte in mobile/lib/ad_pricing.dart — means
// the price the shopper sees is exactly the price PayFast collects.

export type AdPlacement = 'feed' | 'near_me'

export interface AdRateCard {
  currency: 'ZAR'
  perPersonCents: number
  minCents: number
  minReach: number
  maxReach: number
  reachOptions: number[]
  placements: Array<{ id: AdPlacement; label: string; multiplierPct: number }>
  provinces: string[]
}

// R0.08 per person reached, R100 minimum spend. Near-me placement carries a 20%
// premium because it reaches shoppers who are actively standing near a store.
const PER_PERSON_CENTS = 8
const MIN_CENTS = 10_000
const MIN_REACH = 500
const MAX_REACH = 100_000

const PLACEMENTS: AdRateCard['placements'] = [
  { id: 'feed', label: 'Deals feed', multiplierPct: 100 },
  { id: 'near_me', label: 'Near me', multiplierPct: 120 },
]

// South African provinces an advertiser may target. Empty province = national.
export const AD_PROVINCES = [
  'Eastern Cape',
  'Free State',
  'Gauteng',
  'KwaZulu-Natal',
  'Limpopo',
  'Mpumalanga',
  'North West',
  'Northern Cape',
  'Western Cape',
]

export const adRateCard: AdRateCard = {
  currency: 'ZAR',
  maxReach: MAX_REACH,
  minCents: MIN_CENTS,
  minReach: MIN_REACH,
  perPersonCents: PER_PERSON_CENTS,
  placements: PLACEMENTS,
  provinces: AD_PROVINCES,
  reachOptions: [1_000, 2_500, 5_000, 10_000, 25_000],
}

export function isValidAdPlacement(value: unknown): value is AdPlacement {
  return value === 'feed' || value === 'near_me'
}

export function isValidAdProvince(value: unknown): boolean {
  return typeof value === 'string' && AD_PROVINCES.includes(value)
}

export function clampReach(reach: number): number {
  if (!Number.isFinite(reach)) {
    return MIN_REACH
  }

  return Math.min(MAX_REACH, Math.max(MIN_REACH, Math.round(reach)))
}

// The one source of truth for what an ad costs. Always returns an integer number
// of cents that is at least the minimum spend.
export function computeAdPriceCents(input: { reach: number; placement: AdPlacement }): number {
  const reach = clampReach(input.reach)
  const placement = PLACEMENTS.find((option) => option.id === input.placement) ?? PLACEMENTS[0]
  const raw = Math.round((reach * PER_PERSON_CENTS * placement.multiplierPct) / 100)

  return Math.max(MIN_CENTS, raw)
}

export function formatRandFromCents(cents: number): string {
  const amount = cents / 100
  return `R${Number.isInteger(amount) ? amount.toFixed(0) : amount.toFixed(2)}`
}

File: mobile/lib/ad_pricing.dart
// Client mirror of src/services/adPricing.ts. Same formula so the price an
// advertiser sees while choosing reach is exactly what the server charges: the
// server stays authoritative, this only powers the live estimate. The rate card
// itself is fetched from the API and drives the pickers; this file just prices a
// choice locally and formats rand.

import 'widgets/common.dart' show formatMoney;

class AdRateCard {
  const AdRateCard({
    required this.perPersonCents,
    required this.minCents,
    required this.minReach,
    required this.maxReach,
    required this.reachOptions,
    required this.placements,
    required this.provinces,
  });

  final int perPersonCents;
  final int minCents;
  final int minReach;
  final int maxReach;
  final List<int> reachOptions;
  final List<AdPlacementOption> placements;
  final List<String> provinces;

  static const fallback = AdRateCard(
    perPersonCents: 8,
    minCents: 10000,
    minReach: 500,
    maxReach: 100000,
    reachOptions: [1000, 2500, 5000, 10000, 25000],
    placements: [
      AdPlacementOption(id: 'feed', label: 'Deals feed', multiplierPct: 100),
      AdPlacementOption(id: 'near_me', label: 'Near me', multiplierPct: 120),
    ],
    provinces: [
      'Eastern Cape',
      'Free State',
      'Gauteng',
      'KwaZulu-Natal',
      'Limpopo',
      'Mpumalanga',
      'North West',
      'Northern Cape',
      'Western Cape',
    ],
  );

  int clampReach(int reach) {
    if (reach < minReach) return minReach;
    if (reach > maxReach) return maxReach;
    return reach;
  }

  /// The live price estimate for a choice, matching computeAdPriceCents on the
  /// server byte-for-byte.
  int priceCents({required int reach, required String placementId}) {
    final placement = placements.firstWhere(
      (option) => option.id == placementId,
      orElse: () => placements.isNotEmpty
          ? placements.first
          : const AdPlacementOption(id: 'feed', label: 'Deals feed', multiplierPct: 100),
    );
    final clamped = clampReach(reach);
    final raw = (clamped * perPersonCents * placement.multiplierPct / 100).round();
    return raw < minCents ? minCents : raw;
  }

  factory AdRateCard.fromJson(Map<String, dynamic> json) {
    final placements = (json['placements'] as List?)
            ?.whereType<Map>()
            .map((item) =>
                AdPlacementOption.fromJson(Map<String, dynamic>.from(item)))
            .toList() ??
        fallback.placements;
    final reachOptions = (json['reachOptions'] as List?)
            ?.whereType<num>()
            .map((value) => value.toInt())
            .toList() ??
        fallback.reachOptions;
    final provinces = (json['provinces'] as List?)
            ?.whereType<String>()
            .toList() ??
        fallback.provinces;
    return AdRateCard(
      perPersonCents: (json['perPersonCents'] as num?)?.toInt() ?? fallback.perPersonCents,
      minCents: (json['minCents'] as num?)?.toInt() ?? fallback.minCents,
      minReach: (json['minReach'] as num?)?.toInt() ?? fallback.minReach,
      maxReach: (json['maxReach'] as num?)?.toInt() ?? fallback.maxReach,
      reachOptions: reachOptions.isEmpty ? fallback.reachOptions : reachOptions,
      placements: placements.isEmpty ? fallback.placements : placements,
      provinces: provinces.isEmpty ? fallback.provinces : provinces,
    );
  }
}

class AdPlacementOption {
  const AdPlacementOption({
    required this.id,
    required this.label,
    required this.multiplierPct,
  });

  final String id;
  final String label;
  final int multiplierPct;

  factory AdPlacementOption.fromJson(Map<String, dynamic> json) => AdPlacementOption(
        id: json['id']?.toString() ?? 'feed',
        label: json['label']?.toString() ?? 'Deals feed',
        multiplierPct: (json['multiplierPct'] as num?)?.toInt() ?? 100,
      );
}

/// "R2000" for whole rand, "R100.50" when there are cents.
String formatRandFromCents(int cents) => formatMoney(cents);
